Fix and/or accept corrupt attachments.

machisuji opened the pull request:
Also use file column for filename: the uploader uses that one too, anyway
This way things don't break when the file is missing.

Fixes complications with WP [#18106](https://community.openproject.org/work_packages/18106)

machisuji committed:
use file column for filename

the uploader uses that one too, anyway

also this way things don't break when the file is missing

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://fix-attachments-filename.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://fix-attachments-filename.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

machisuji committed:
restore original disk filename if available

machisuji committed:
set file column for corrupt attachments

i.e. attachments whose file could not be found in the original attachments  migration

machisuji committed:
fixed spec

linki commented on app/models/attachment.rb:
```diff
@@ -139,7 +139,7 @@ def diskfile
   end
 
   def filename
-    file.file.filename if file.file
```
:scream:

NobodysNightmare replied:
:+1:

ulferts commented on db/migrate/20150116095004_patch_corrupt_attachments.rb:
```diff
@@ -0,0 +1,49 @@
+  def down
+    puts "Can't revert this migration as it would mean breaking valid attachments. \
+          Even if we wanted to do that we can't know which ones were broken before \
+          to only break those again.".squish
```
Won't the correct message here state that it is possible to revert the migrations but it is not intended to change anything as it would only corrupt things?

machisuji replied:
Table before the migration (after the first one that skipped corrupt attachments):

| filename | diskfilename | file |
| --- | --- | --- |
| found.jpg | 20131107_found.jpg | found.jpg |
| not_found.gif | 20140503_not_found.gif |  |

And after this migration:

| filename | diskfilename | file |
| --- | --- | --- |
| found.jpg | 20131107_found.jpg | found.jpg |
| not_found.gif | 20140503_not_found.gif | not_found.gif |

So from the data in the table alone we can not deduct whether or not the attachment was broken before.
We could, however, check if the file can be found and if not we just delete the file column and we would have
the same state as before. But yes, that would corrupt things even more. I can adjust the error message with this bit.

machisuji committed:
You can't or you won't?
